electrumsv/wallet_database.py

```diff
@@ -1757,6 +1764,249 @@ def delete_reorged_entries(self, reorg_height: int) -> None:
+        return self.utxos.__iter__()
+
+    def __getitem__(self):
+        return self.utxos.__getitem__()
+
+    def __setitem__(self):
+        return self.utxos.__setitem__()
+
+    def __delitem__(self):
+        return self.utxos.__delitem__()
+
+    def extract_utxos_from_tx_entry(self, tx_cache_entry: TxCacheEntry) -> List[UTXO]:
+        """Takes a StateCleared transaction and transforms the outputs to a list of UTXOs"""
+        # TODO - think about thread lock requirements
+        new_utxos = []
+        for index, output in enumerate(tx_cache_entry.transaction.outputs):
+
+            if output.script_pubkey.to_asm().find('OP_RETURN') != -1:
+                continue
+
+            new_utxo = UTXO(value=output.value,
+                            script_pubkey=output.script_pubkey,
+                            tx_hash=tx_cache_entry.transaction.txid(),
+                            out_index=index,
+                            height=tx_cache_entry.metadata.height,
+                            address=classify_output_script(output.script_pubkey, Net.COIN),
+                            is_coinbase=tx_cache_entry.transaction.is_coinbase())
+            self.logger.debug(f"extracted utxo: {new_utxo}")
+            new_utxos.append(new_utxo)
+
+        return new_utxos
+
+    def add_from_tx_entry(self, tx_cache_entry: TxCacheEntry) -> None:
+        """This should only be called after tx reaches StateCleared
+        and only for incoming payments. (Called by handle_incoming_payments)"""
+        assert(tx_cache_entry.flags | TxFlags.StateCleared), "Only StateCleared transactions " \
+                                                             "should be added to utxo cache."
+        # re-calculates tx object from bytedata but simplifies function signatures
+        tx = tx_cache_entry.transaction
+        self.logger.debug(f"updating utxo cache for {tx.txid()}")
+        with self._lock:
+            # Add new utxos to cache
+            new_utxos = self.extract_utxos_from_tx_entry(tx_cache_entry)
+            self.utxos.extend(new_utxos)
+
+    def undo_add_from_tx_entry(self, tx_cache_entry: TxCacheEntry) -> None:
+        """In the case of a chain reorg in which transaction regresses from
+        StateCleared --> StateDispatched, we must 'undo' the addition to cache."""
+        with self._lock:
+            raise NotImplementedError
+
+    def remove_utxos_by_key(self, keys: Set[Tuple[str, int]]) -> None:
+        with self._lock:
+            for key in keys:
+                for utxo in self.utxos:
+                    if key == utxo.key():
+                        self.utxos.remove(utxo)
+                        break
+
+    def remove_utxos(self, utxos_for_removal: Optional[List[UTXO], UTXOCache]) -> None:
+        """Removes frozen utxos from cache"""
+        with self._lock:
+            matched_utxos = []
+            for frozen in utxos_for_removal:
+                for index, utxo in enumerate(self.utxos):
```

Comment:
Actually there is a reason for doing it that way.
Deque.remove() is O(n) (I tested it - it's bad) so I compensate here by only iterating in from the left hand side of the deque until it hits a match - if deque is maintained in order with popping used utxos from left and appending new utxos to right side then a match should be found in the first few items... 

The reason I am now contemplating specifically: **OrderedDict** now is because (I only just learned of this)...
OrderedDict has a "move_to_end" method which is O(1) and when combined with O(1) getitem, setitem, delitem, and update... **and** the fact that it is ordered... I see no downsides to it! (I might just use OrderedDict as the only datatype I ever use! :rofl: )

The whole reason I went for deque was because I envisioned (ab)using the O(1) pop and appends from both ends and maintaining some kind of sort for high throughput...

The hardest part too ^^ work has been getting the sequence of events right. Swapping out the underlying datatype for utxo cache will no be so difficult.